### *“Success is no accident. It is hard work, perseverance, learning, studying, sacrifice and most of all, love of what you are doing or learning to do.”* 

*-- Pele, Brazilian soccer player*

You're doing well and that's not  by accident.  So continue to work hard!  Keep up the good work!

----

### YOUR NAME HERE:  Travis Kerr

#### List the names of anyone you helped here: Brendan, Evan

#### List the names of anyone who helped you here: Brendan, Evan
----

# HW 2: GROUP BY, HAVING

**Be sure to use your SQL cheat sheet to help you in this homework!**

The goal of this homework will be to practice GROUP BY and HAVING, plus review other concepts that you know.

Recall that the correct order of a SQL query is below.

```
SELECT column_name(s)
FROM table_name
WHERE condition                     
GROUP BY column_name(s)
HAVING condition
ORDER BY column_name(s);
```


---

# **RECALL:**  Points will be deducted for not adhering to conventions.  

So remember to:

* Capitalize SQL keywords.
* Line up the main keywords in a column (SELECT/FROM/WHERE/GROUP BY/HAVING/ORDER BY).
* Indent when using multiple lines in the SELECT or WHERE clause.
* Line up Case statements nicely.

---

# How to Read in from a SQLite DB when you are within a notebook File

To use SQLite in a notebook, we need to import some special packages.

Read through the comments in the Python code below  and get a feel of what is happening.

Note that comments begin with a hashtag in Python.

!!!!!!!!!!!!!!!!!!!!!!So please run the cells below!!!!!!!!!!!!!!!!!!!!

In [1]:
#This cell imports the SQLite package, plus 2 other packages 
#that allow us to manipulate tables of data within a Python notebook file.

import pandas as pd     
import numpy as np    
import sqlite3 as sql     

In the next cell, we connect to the ```hearts_on_fire``` database.

In [2]:
#Try to make a connection to a database file ('hearts_on_fire.db'). 
#If anything goes wrong, print an error message.

try:
    conn = sql.connect('hearts_on_fire.db')
except ValueError: 
    print("""Dr. R Note: A ValueError occurred:\n\t""" + ValueError )

# Recall what's in the Hearts on Fire Database. 

Yet again, we will be using the ```hearts_on_fire``` data.  

Recall that Hearts on Fire is a campaign that DSU runs every year to fundraise donations.

Here's what each of the columns mean.

* donation:  the amount the contributor contributed
* state:  where the alumnus lives
* affiliation: the status of the donor at the time of the donation
  * Indicated if the contributor was an alumni, student, family member, etc., when the donation was made.
* class_year:  the year the person graduated.
* preferred_division:  where the contributor wishes to send their funds.
* date_submitted: the date and time of the contributor made the donation
* donation year: The year of the hearts on fire campaign

Read through the comments in the Python code below and get a feel of what is happening.

In [3]:
#Create a sql statement between the triple quotes.
sql_statement_all = """
                       SELECT *
                       FROM hearts_on_fire;
                       
                    """

#Now run the query, passing it and the connection (conn) into the needed Python method.
#You get a dataframe (Python version of a table) back with the results of the query.

query_results = pd.read_sql_query(sql_statement_all, conn)

#Print the dataframe results.
query_results

id  donation state affiliation  class_year   preferred_division  \
0        1      50.0    CA      Alumni      1997.0      Performing Arts   
1        2      37.0    PA     Student      2020.0             Business   
2        3      25.0    PA      Alumni         NaN                 None   
3        4      10.0    PA     Student      2022.0         Science/Math   
4        5      50.0    PA      Family         NaN                 None   
...    ...       ...   ...         ...         ...                  ...   
1779  1780     100.0    NY      Alumni      2019.0  Liberal Arts/Social   
1780  1781      20.0    FL      Family      2023.0                 None   
1781  1782     100.0    WV      Alumni      1970.0         Science/Math   
1782  1783     250.0    DE      Alumni      2007.0  Liberal Arts/Social   
1783  1784    2500.0    CT      Alumni      1971.0                 None   

       date_submitted  donation_year  
0     1/29/2020 10:57           2020  
1     2/12/2020 21:31           2020  
2     2/10/2020 15:10           2020  
3     2/12/2020 17:53           2020  
4     2/12/2020 15:47           2020  
...               ...            ...  
1779   2/8/2021 11:53           2021  
1780   2/9/2021 11:44           2021  
1781  2/10/2021 15:01           2021  
1782  2/10/2021 18:23           2021  
1783  2/10/2021 13:09           2021  

[1784 rows x 8 columns]

## How NULLs work when running within a Python notebook

You know from your last HW that the ```hearts_on_fire``` database has some NULL values within it. 

In fact, you discovered that the hearts_on_fire data has:

* 3 NULLs in the state column
* 683 NULLs in the class_year column
* 936 NULLs in the preferred_division column.

However, when looking at the data within a Jupyter notebook file, you won't see the word NULL, even though there are NULLS in the database.

Here's what will happen.  We use Python to request rows from the database.  When the records are returned, the notebook file will render/show the NULLs differently that they were shown in the database project.

So NULLs will instead appear in a Jupyter notebook as one of these values.

* NaN (Not a number)
* None

However, the SQL commands ```IS NULL``` and ```IS NOT NULL``` are still usable and still work just fine.  Run the next cell to prove this to yourself.


In [4]:
#Here we put a SQL query between the triple quotes.
sql_statement_nulls = """ 
                          SELECT 1784 - COUNT(donation) AS missing_dontations, 
                                 1784 - COUNT(state) AS missing_states, 
                                 1784 - COUNT(affiliation) AS missing_affiliations, 
                                 1784 - COUNT(class_year) AS missing_class_years, 
                                 1784 - COUNT(preferred_division) AS missing_divisions, 
                                 1784 - COUNT(date_submitted) AS missing_submission_dates, 
                                 1784 - COUNT(donation_year) AS missing_donation_years
                            FROM hearts_on_fire;
                      """

#Run the query, passing it and the connection (conn) into the needed Python method.
#You get a dataframe (Python version of a table) back with the results of the query.
query_results = pd.read_sql_query(sql_statement_nulls, conn)

#Print the dataframe results.
query_results

missing_dontations  missing_states  missing_affiliations  \
0                   0               3                     0   

   missing_class_years  missing_divisions  missing_submission_dates  \
0                  683                936                         0   

   missing_donation_years  
0                       0

---

# QUESTION 1

Replace the ellipsis (...) in the triple quotes below with a SQL query that retrieves the 3 records where a state is missing.

Show all columns for these records.

***Notice that the states report as ```None``` instead of ```NULL```.***

***Please, please indent your queries nicely.***


In [ ]:
#Replace the ellipsis (...) with a SELECT statement.
sql_statement1 = """
                    SELECT *
                    FROM hearts_on_fire
                    WHERE state IS NULL;
                 """

#Run the query and save the returned dataframe.
query_results1 = pd.read_sql_query(sql_statement1, conn)

#Print the dataframe results.
query_results1

id  donation state affiliation class_year preferred_division  \
0  266     100.0  None    Employee       None               None   
1  441     500.0  None    Employee       None               None   
2  632      50.0  None      Family       None               None   

   date_submitted  donation_year  
0  2/12/2020 9:30           2020  
1  2/12/2020 8:07           2020  
2  2/7/2020 16:24           2020

# QUESTION 2

Show the year and the largest/smallest/mean donation amount for that year, plus how many contributors there were for that year.

The column names should be: 

* ```donation_year, num_donors, max_donation, min_donation, avg_donation```.



Double-click __here__ to see what the right output looks like.

<!-- 

| donation_year | num_donors | max_donation | min_donation | avg_donation |
|--------------:|-----------:|-------------:|--------------:|-------------:|
| 2020          | 855        | 10000.0      | 1.0           | 139.423485   |
| 2021          | 929        | 10000.0      | 5.0           | 176.956835   |

-->

In [ ]:
sql_statement2 = """
                    SELECT donation_year,
                           COUNT(*) AS num_donors,
                           MAX(donation) AS max_donation,
                           MIN(donation) AS min_donation,
                           AVG(donation) AS avg_donation
                    FROM hearts_on_fire
                    GROUP BY donation_year
                 
                 """

#Run the query and save the returned dataframe.
query_results2 = pd.read_sql_query(sql_statement2, conn)

#Print the dataframe results.
query_results2

donation_year  num_donors  max_donation  min_donation  avg_donation
0           2020         855       10000.0           1.0    139.423485
1           2021         929       10000.0           5.0    176.956835

# QUESTION 3

Determine how many contributors there were per state.

Order your result from the state with the highest number of contributors to the lowest.  

Rename the column with the number of contributors to appear as ``num_contributors``.



Double-click __here__ to see what the right output looks like.

<!-- 

| state | num_contributors |
|------:|-----------------:|
| PA    | 1199             |
| NJ    | 193              |
| NY    | 104              |
| DE    | 47               |
| MD    | 39               |
| CA    | 27               |
| FL    | 16               |
| CT    | 16               |
| VA    | 14               |
| OH    | 12               |
| NC    | 12               |
| IL    | 11               |
| CO    | 11               |
| TX    | 8                |
| MA    | 8                |
| MI    | 7                |
| TN    | 5                |
| SC    | 5                |
| NH    | 5                |
| DC    | 5                |
| GA    | 4                |
| WA    | 3                |
| RI    | 3                |
| OR    | 3                |
| NV    | 3                |
| None  | 3                |
| WV    | 2                |
| OK    | 2                |
| NM    | 2                |
| MO    | 2                |
| ME    | 2                |
| IN    | 2                |
| AZ    | 2                |
| UT    | 1                |
| MT    | 1                |
| MN    | 1                |
| LA    | 1                |
| KY    | 1                |
| HI    | 1                |
| AK    | 1                |

-->

In [ ]:
sql_statement3 = """
                    SELECT state, COUNT(*) AS num_contributors
                    FROM hearts_on_fire
                    GROUP by state
                    ORDER BY num_contributors DESC;
                 """

#Run the query and save the returned dataframe.
query_results3 = pd.read_sql_query(sql_statement3, conn)

#Print the dataframe results.
query_results3

state  num_contributors
0     PA              1199
1     NJ               193
2     NY               104
3     DE                47
4     MD                39
5     CA                27
6     FL                16
7     CT                16
8     VA                14
9     OH                12
10    NC                12
11    IL                11
12    CO                11
13    TX                 8
14    MA                 8
15    MI                 7
16    TN                 5
17    SC                 5
18    NH                 5
19    DC                 5
20    GA                 4
21    WA                 3
22    RI                 3
23    OR                 3
24    NV                 3
25  None                 3
26    WV                 2
27    OK                 2
28    NM                 2
29    MO                 2
30    ME                 2
31    IN                 2
32    AZ                 2
33    UT                 1
34    MT                 1
35    MN                 1
36    LA                 1
37    KY                 1
38    HI                 1
39    AK                 1

# QUESTION 4

That was a lot of rows!  

Redo the previous question so that only the top 10 states (in terms of number of contributors) show.  


Double-click __here__ to see what the right output looks like.

<!--

| state | num_contributors |
|------:|-----------------:|
| PA    | 1199             |
| NJ    | 193              |
| NY    | 104              |
| DE    | 47               |
| MD    | 39               |
| CA    | 27               |
| FL    | 16               |
| CT    | 16               |
| VA    | 14               |
| OH    | 12               |

-->

In [ ]:

sql_statement4 = """
                    SELECT state, COUNT(*) AS num_contributors
                    FROM hearts_on_fire
                    GROUP by state
                    ORDER BY num_contributors DESC
                    LIMIT 10;
                 """

#Run the query and save the returned dataframe.
query_results4 = pd.read_sql_query(sql_statement4, conn)

#Print the dataframe results.
query_results4

state  num_contributors
0    PA              1199
1    NJ               193
2    NY               104
3    DE                47
4    MD                39
5    CA                27
6    FL                16
7    CT                16
8    VA                14
9    OH                12

# QUESTION 5

Recall that the Northeast Region is CT, ME, MA, NH, RI, VT, NJ, NY, PA.
  
Write a query to determine which states OUTSIDE of the Northeast Region had 1 of the 5 highest amounts of alumni contributors in the year 2021.

Your columns should be pertinent to the question at hand.


Double-click __here__ to see what the right output looks like.

<!--

| state | num_alumni |
|------:|-----------:|
| DE    | 17         |
| CA    | 10         |
| MD    | 9          |
| NC    | 6          |
| IL    | 5          |

-->

In [ ]:

sql_statement5 = """
                    SELECT state, COUNT(*) AS num_alumni
                    FROM hearts_on_fire
                    WHERE state NOT IN("CT", "ME", "MA", "NH", "RI", "VT", "NJ", "NY", "PA") AND
                          affiliation = "Alumni" AND
                          donation_year = 2021
                    GROUP BY state
                    ORDER BY num_alumni DESC
                    LIMIT 5;
                 """

#Run the query and save the returned dataframe.
query_results5 = pd.read_sql_query(sql_statement5, conn)

#Print the dataframe results.
query_results5

state  num_alumni
0    DE          17
1    CA          10
2    MD           9
3    NC           6
4    IL           5

# QUESTION 6

Show the average donation (```avg_donation```) for each of the states in the NE region.


Double-click __here__ to see what the right output looks like.

<!--

| state | avg_donation |
|------:|-------------:|
| CT    | 225.312500   |
| MA    | 656.250000   |
| ME    | 275.000000   |
| NH    | 66.000000    |
| NJ    | 285.253886   |
| NY    | 92.423077    |
| PA    | 131.153445   |
| RI    | 45.000000    |

-->

In [ ]:

sql_statement6 = """
                    SELECT state, AVG(donation)
                    FROM hearts_on_fire
                    WHERE state IN ("CT", "ME", "MA", "NH", "RI", "VT", "NJ", "NY", "PA")
                    GROUP BY state;

                 """

#Run the query and save the returned dataframe.
query_results6 = pd.read_sql_query(sql_statement6, conn)

#Print the dataframe results.
query_results6

state  AVG(donation)
0    CT     225.312500
1    MA     656.250000
2    ME     275.000000
3    NH      66.000000
4    NJ     285.253886
5    NY      92.423077
6    PA     131.153445
7    RI      45.000000

# QUESTION 7

For the year 2020, show the number of donors per the amount of donation.  Use ```num_donors``` as a column name.

In the final results, order the donations from smallest to largest.

Double-click __here__ to see what the right output looks like.

<!--

| donation | num_donors |
|---------:|------------|
| 1.00     | 3          |
| 2.00     | 3          |
| 3.00     | 1          |
| 5.00     | 45         |
| 5.98     | 1          |
| 6.00     | 1          |
| 10.00    | 62         |
| 15.00    | 15         |
| 20.00    | 55         |
| 20.08    | 1          |
| 20.16    | 1          |
| 20.17    | 1          |
| 25.00    | 185        |
| 30.00    | 24         |
| 35.00    | 2          |
| 36.00    | 1          |
| 37.00    | 2          |
| 40.00    | 10         |
| 45.00    | 1          |
| 50.00    | 124        |
| 55.00    | 1          |
| 60.00    | 2          |
| 67.00    | 1          |
| 72.00    | 1          |
| 75.00    | 7          |
| 100.00   | 169        |
| 100.19   | 1          |
| 120.00   | 1          |
| 125.00   | 4          |
| 150.00   | 21         |
| 200.00   | 15         |
| 225.00   | 1          |
| 250.00   | 29         |
| 300.00   | 7          |
| 400.00   | 3          |
| 500.00   | 32         |
| 750.00   | 1          |
| 800.00   | 1          |
| 1000.00  | 11         |
| 1500.00  | 1          |
| 2000.00  | 1          |
| 2500.00  | 2          |
| 3000.00  | 1          |
| 4348.50  | 1          |
| 5000.00  | 1          |
| 10000.00 | 2          |

-->

In [ ]:

sql_statement7 = """
                    SELECT donation, COUNT(*) AS num_donors
                    FROM hearts_on_fire
                    WHERE donation_year = 2020
                    GROUP BY donation
                    ORDER BY donation ASC;
                 """

query_results7 = pd.read_sql_query(sql_statement7, conn)
query_results7

donation  num_donors
0       1.00           3
1       2.00           3
2       3.00           1
3       5.00          45
4       5.98           1
5       6.00           1
6      10.00          62
7      15.00          15
8      20.00          55
9      20.08           1
10     20.16           1
11     20.17           1
12     25.00         185
13     30.00          24
14     35.00           2
15     36.00           1
16     37.00           2
17     40.00          10
18     45.00           1
19     50.00         124
20     55.00           1
21     60.00           2
22     67.00           1
23     72.00           1
24     75.00           7
25    100.00         169
26    100.19           1
27    120.00           1
28    125.00           4
29    150.00          21
30    200.00          15
31    225.00           1
32    250.00          29
33    300.00           7
34    400.00           3
35    500.00          32
36    750.00           1
37    800.00           1
38   1000.00          11
39   1500.00           1
40   2000.00           1
41   2500.00           2
42   3000.00           1
43   4348.50           1
44   5000.00           1
45  10000.00           2

# QUESTION 8

There were a lot of 1s in the prior table.  

Run the same query but this time, show only those rows where more than 1 person contributed the given amount.

Double-click __here__ to see what the right output looks like.

<!--


-->

In [ ]:

sql_statement8 = """
                    SELECT donation, COUNT(*) AS num_donors
                    FROM hearts_on_fire
                    WHERE donation_year = 2020
                    GROUP BY donation
                    HAVING num_donors > 1
                    ORDER BY donation ASC;

                 """

query_results8 = pd.read_sql_query(sql_statement8, conn)
query_results8

donation  num_donors
0        1.0           3
1        2.0           3
2        5.0          45
3       10.0          62
4       15.0          15
5       20.0          55
6       25.0         185
7       30.0          24
8       35.0           2
9       37.0           2
10      40.0          10
11      50.0         124
12      60.0           2
13      75.0           7
14     100.0         169
15     125.0           4
16     150.0          21
17     200.0          15
18     250.0          29
19     300.0           7
20     400.0           3
21     500.0          32
22    1000.0          11
23    2500.0           2
24   10000.0           2

# QUESTION 9

Show how many contributors there were per donation year/affiliation pairing. Call one column ```num_donors```.

Order your table so that the donation years increase while the affiliations are in alphabetical order per year.

Double-click __here__ to see what the right output looks like.

<!--
| donation_year | affiliation | num_donors |
|--------------:|------------:|-----------:|
| 2020          | Alumni      | 470        |
| 2020          | Employee    | 175        |
| 2020          | Family      | 128        |
| 2020          | Friend      | 31         |
| 2020          | Student     | 51         |
| 2021          | Alumni      | 466        |
| 2021          | Employee    | 167        |
| 2021          | Family      | 150        |
| 2021          | Friend      | 61         |
| 2021          | Student     | 85         |

-->

In [ ]:

sql_statement9 = """
                    SELECT donation_year, affiliation, COUNT(*) AS num_donors
                    FROM hearts_on_fire
                    GROUP BY donation_year, affiliation;
                 """

query_results9 = pd.read_sql_query(sql_statement9, conn)
query_results9

donation_year affiliation  num_donors
0           2020      Alumni         470
1           2020    Employee         175
2           2020      Family         128
3           2020      Friend          31
4           2020     Student          51
5           2021      Alumni         466
6           2021    Employee         167
7           2021      Family         150
8           2021      Friend          61
9           2021     Student          85

# QUESTION 10

Show the average donations per reported class year (1969, 1970, 1980, etc), ordered from highest average donation to lowest.  

Show the class year against an   ```avg_donation``` column.

Double-click __here__ to see what the right output looks like.

<!--

| class_year | avg_donation |
|-----------:|-------------:|
| 1994.0     | 1139.090909  |
| 1971.0     | 1129.545455  |
| 1985.0     | 899.230769   |
| 1969.0     | 788.690476   |
| 1990.0     | 746.153846   |
| 1970.0     | 576.562500   |
| 1997.0     | 538.636364   |
| 1979.0     | 530.000000   |
| 1984.0     | 335.000000   |
| 1977.0     | 300.000000   |
| 1976.0     | 275.000000   |
| 1981.0     | 237.500000   |
| 2007.0     | 220.685185   |
| 1978.0     | 216.666667   |
| 2000.0     | 216.428571   |
| 1982.0     | 208.571429   |
| 1980.0     | 195.000000   |
| 1983.0     | 185.000000   |
| 1998.0     | 181.491071   |
| 1987.0     | 167.857143   |
| 1991.0     | 165.000000   |
| 1996.0     | 156.111111   |
| 1975.0     | 153.333333   |
| 1973.0     | 140.666667   |
| 2003.0     | 129.166667   |
| NaN        | 127.705959   |
| 1993.0     | 125.000000   |
| 2009.0     | 123.500000   |
| 1988.0     | 115.625000   |
| 1986.0     | 113.333333   |
| 2002.0     | 110.333333   |
| 2012.0     | 102.777778   |
| 2004.0     | 100.227273   |
| 1974.0     | 100.000000   |
| 2006.0     | 91.750000    |
| 1995.0     | 89.125000    |
| 2005.0     | 86.739130    |
| 2014.0     | 83.888889    |
| 1992.0     | 83.333333    |
| 2001.0     | 81.250000    |
| 2023.0     | 76.921053    |
| 2019.0     | 72.265625    |
| 1999.0     | 69.783784    |
| 2015.0     | 65.441176    |
| 2016.0     | 65.222500    |
| 2010.0     | 64.341250    |
| 1972.0     | 62.500000    |
| 2017.0     | 62.068966    |
| 2020.0     | 58.843243    |
| 1989.0     | 58.333333    |
| 2013.0     | 55.666667    |
| 2008.0     | 54.226667    |
| 2011.0     | 52.407407    |
| 2021.0     | 51.296296    |
| 2025.0     | 50.666667    |
| 2022.0     | 50.000000    |
| 2024.0     | 43.804348    |
| 2018.0     | 31.621622    |
-->

In [ ]:

sql_statement10 = """ 
                     SELECT class_year, AVG(donation) AS avg_donation
                     FROM hearts_on_fire
                     GROUP BY class_year
                     ORDER BY avg_donation DESC;
                  """

query_results10 = pd.read_sql_query(sql_statement10, conn)
query_results10

class_year  avg_donation
0       1994.0   1139.090909
1       1971.0   1129.545455
2       1985.0    899.230769
3       1969.0    788.690476
4       1990.0    746.153846
5       1970.0    576.562500
6       1997.0    538.636364
7       1979.0    530.000000
8       1984.0    335.000000
9       1977.0    300.000000
10      1976.0    275.000000
11      1981.0    237.500000
12      2007.0    220.685185
13      1978.0    216.666667
14      2000.0    216.428571
15      1982.0    208.571429
16      1980.0    195.000000
17      1983.0    185.000000
18      1998.0    181.491071
19      1987.0    167.857143
20      1991.0    165.000000
21      1996.0    156.111111
22      1975.0    153.333333
23      1973.0    140.666667
24      2003.0    129.166667
25         NaN    127.705959
26      1993.0    125.000000
27      2009.0    123.500000
28      1988.0    115.625000
29      1986.0    113.333333
30      2002.0    110.333333
31      2012.0    102.777778
32      2004.0    100.227273
33      1974.0    100.000000
34      2006.0     91.750000
35      1995.0     89.125000
36      2005.0     86.739130
37      2014.0     83.888889
38      1992.0     83.333333
39      2001.0     81.250000
40      2023.0     76.921053
41      2019.0     72.265625
42      1999.0     69.783784
43      2015.0     65.441176
44      2016.0     65.222500
45      2010.0     64.341250
46      1972.0     62.500000
47      2017.0     62.068966
48      2020.0     58.843243
49      1989.0     58.333333
50      2013.0     55.666667
51      2008.0     54.226667
52      2011.0     52.407407
53      2021.0     51.296296
54      2025.0     50.666667
55      2022.0     50.000000
56      2024.0     43.804348
57      2018.0     31.621622

# QUESTION 11

Notice that in your previous question, there was a category for NaN where the average was about 127.71.  

If there was not, then your answer is incorrect and please correct it.

Write an english sentence for what you learn from this row.  The value $127.71 should be in your answer.

*Write your answer here: This is the average donation of the people who donated without filling out their class year, making this a NULL category. 

# QUESTION 12

DeSales was founded in 1962.  

Determine how many donors and the average donation per decade (1960s, 1970s, etc).  

Rename the column headers ```as num_donors``` and ```avg_donation```.

For those who did not report a class year, mark them as ```No class year reported```.



Double-click __here__ to see what the right output looks like.

<!--

|                 decade | num_donors | avg_donation |
|-----------------------:|-----------:|--------------|
| 1960s                  | 42         | 788.690476   |
| 1970s                  | 55         | 510.309091   |
| 1980s                  | 80         | 307.125000   |
| 1990s                  | 168        | 293.629464   |
| 2000s                  | 246        | 122.238455   |
| 2010s                  | 326        | 64.147853    |
| 2020s                  | 184        | 55.990217    |
| No class year reported | 683        | 127.705959   |

-->

In [22]:
sql_statement12 = """
                    SELECT COUNT(*) AS num_donors, AVG(donation) AS avg_donation,
                      CASE
                                WHEN class_year >= 1960 AND class_year <= 1969 THEN "1960s"
                                WHEN class_year >= 1970 AND class_year <= 1979 THEN "1970s"
                                WHEN class_year >= 1980 AND class_year <= 1989 THEN "1980s"
                                WHEN class_year >= 1990 AND class_year <= 1999 THEN "1990s"
                                WHEN class_year >= 2000 AND class_year <= 2009 THEN "2000s"
                                WHEN class_year >= 2010 AND class_year <= 2019 THEN "2010s"
                                WHEN class_year >= 2020 AND class_year <= 2029 THEN "2020s"
                                ELSE "No class year reported"
                      END decade
                    FROM hearts_on_fire
                    GROUP by decade
                  """


query_results12 = pd.read_sql_query(sql_statement12, conn)
query_results12

num_donors  avg_donation                  decade
0          42    788.690476                   1960s
1          55    510.309091                   1970s
2          80    307.125000                   1980s
3         168    293.629464                   1990s
4         246    122.238455                   2000s
5         326     64.147853                   2010s
6         184     55.990217                   2020s
7         683    127.705959  No class year reported

# QUESTION 13

We'll say a high-donor is one who contributes 500 or more dollars.  

Show the number of high-donors in each state.

Order your results from the largest number of high donors to the smallest.

Also, name the column with the number of high donors as: ``num_high_donors``.


Double-click __here__ to see what the right output looks like.

<!--

| state | num_high_donors |
|------:|----------------:|
| PA    | 75              |
| NJ    | 21              |
| DE    | 6               |
| MD    | 5               |
| MA    | 3               |
| VA    | 2               |
| NY    | 2               |
| MO    | 2               |
| MI    | 2               |
| CT    | 2               |
| CA    | 2               |
| WA    | 1               |
| OH    | 1               |
| NM    | 1               |
| NC    | 1               |
| ME    | 1               |
| IL    | 1               |
| CO    | 1               |
| None  |                 |

-->

In [ ]:
sql_statement13 = """
                    SELECT state, COUNT(*) AS num_high_donors
                    FROM hearts_on_fire
                    WHERE donation >= 500
                    GROUP BY state
                    ORDER BY num_high_donors DESC;
                   """


query_results13 = pd.read_sql_query(sql_statement13, conn)
query_results13

state  num_high_donors
0     PA               75
1     NJ               21
2     DE                6
3     MD                5
4     MA                3
5     VA                2
6     NY                2
7     MO                2
8     MI                2
9     CT                2
10    CA                2
11    WA                1
12    OH                1
13    NM                1
14    NC                1
15    ME                1
16    IL                1
17    CO                1
18  None                1

# QUESTION 14

In one table, show how many donors fall into the following categories for 2021.

NOTE: Hard brackets indicate inclusion whereas parentheses indicate exclusion.

* [1, 50)
* [50, 100)
* [100,200)
* [200,300)
* [300,400)
* [400,500)
* [500,1000)
* [1000,2000)
* [2000, infinity)


Double-click __here__ to see what the right output looks like.

<!--

|   donation_level | num_donors |
|-----------------:|-----------:|
| [1, 50)          | 321        |
| [100,200)        | 244        |
| [1000,2000)      | 19         |
| [200,300)        | 95         |
| [2000, infinity) | 12         |
| [300,400)        | 10         |
| [400,500)        | 4          |
| [50, 100)        | 179        |
| [500,1000)       | 45         |
   
-->

In [24]:

sql_statement14 = """
                    SELECT COUNT(*),
                      CASE 
                          WHEN donation BETWEEN 1 AND 49 THEN "[1,50)"
                          WHEN donation BETWEEN 50 AND 99 THEN "[50,99)"
                          WHEN donation BETWEEN 100 AND 199 THEN "[100,199)"
                          WHEN donation BETWEEN 200 AND 299 THEN "[200,299)"
                          WHEN donation BETWEEN 300 AND 399 THEN "[300,399)"
                          WHEN donation BETWEEN 400 AND 499 THEN "[400,499)"
                          WHEN donation BETWEEN 500 AND 999 THEN "[500,999)"
                          WHEN donation BETWEEN 1000 AND 1999 THEN "[1000,2000)"
                          WHEN donation >= 2000 THEN "[2000, infinity)"
                      END donation_level
                    FROM hearts_on_fire
                    WHERE donation_year = 2021
                    GROUP BY donation_level
                  """


query_results14 = pd.read_sql_query(sql_statement14, conn)
query_results14

COUNT(*)    donation_level
0       321            [1,50)
1       244         [100,199)
2        19       [1000,2000)
3        95         [200,299)
4        12  [2000, infinity)
5        10         [300,399)
6         4         [400,499)
7       179           [50,99)
8        45         [500,999)

# QUESTION 15

Here's a question you answered with 4 queries in the previous HW. 

Now answer it using just one query.

The US Census lists the following as the 4 main regions of the US.
* Northeast Region
  * CT, ME, MA, NH, RI, VT, NJ, NY, PA
* Midwest Region
  * IL, IN, MI, OH, WI, IA, KS, MN, MO, NE, ND, SD
* South Region
  * DE, DC, FL, GA, MD, NC, SC, VA, WV, AL, KY, MS, TN, AR, LA, OK, TX
* West Region
  * AZ, CO, ID, MT, NV, NM, UT, WY, AK, CA, HI, OR, WA

Write ONE query to tell us what % of contributors came from each region.  

Call the columns ```region``` and ```percentage_donors```.



Double-click __here__ to see what the right output looks like.

<!--

|           region | percentage_donors |
|-----------------:|-------------------|
| Northeast Region | 85.762332         |
| West Region      | 9.024664          |
| South Region     | 3.082960          |
| Midwest Region   | 1.961883          |
| None             | 0.168161          |

-->

In [29]:
sql_statement15 = """
                     SELECT COUNT(*) / 1784.0 * 100 AS percentage_donors,
                       CASE
                         WHEN state IN ("CT", "ME", "MA", "NH", "RI", "VT", "NJ", "NY", "PA") THEN "Northeast Region"
                         WHEN state IN ("IL", "IN", "MI", "OH", "WI", "IA", "KS", "MN", "MO", "NE", "ND", "SD") THEN "Midwest Region"
                         WHEN state IN ("DE", "DC", "FL", "GA", "MD", "NC", "SC", "VA", "WV", "AL", "KY", "MS", "TN", "AR", "LA", "OK", "TX") THEN "South Region"
                         WHEN state IN ("AZ", "CO", "ID", "MT", "NV", "NM", "UT", "WY", "AK", "CA", "HI", "OR", "WA") THEN "West Region"
                         ELSE "None"
                       END region
                     FROM hearts_on_fire
                     GROUP BY region
                     ORDER BY percentage_donors DESC
                  """


query_results15 = pd.read_sql_query(sql_statement15, conn)
query_results15

percentage_donors            region
0          85.762332  Northeast Region
1           9.024664      South Region
2           3.082960       West Region
3           1.961883    Midwest Region
4           0.168161              None

# QUESTION 16

Ask an interesting question that was not already asked, one that uses GROUP BY.  

Make it interesting and informative, ie, don't give me a lame question to just get points.

*What is the amount of students who donated 25 or more from each state in the Northeast Region and graduate during 2024?

In [21]:
sql_statement16 = """
                    SELECT state, COUNT(*) AS num_donors
                    FROM hearts_on_fire
                    WHERE state IN ("CT", "ME", "MA", "NH", "RI", "VT", "NJ", "NY", "PA") AND
                          affiliation = 'Student' AND
                          class_year = 2024 AND
                          donation >= 25
                    GROUP BY state
                  """

query_results16 = pd.read_sql_query(sql_statement16, conn)
query_results16

state  num_donors
0    NJ           1
1    NY           1
2    PA           1

# QUESTION 17

The following query fails in DataCamp.   Why?-  What's Wrong?

``` 
SELECT country, AVG(gross)
FROM films
WHERE AVG(gross) > 1000000
GROUP BY country
```

*Write Your answer here:  There is an aggregate trying to be used in the WHERE section*

# QUESTION 18

Did you follow the conventions expected?  If not, puts will be taken away.

----

## Closing the DB connection.

Throughout this whole file, we've kept our connection to the Database (the hearts_on_fire.db file) open.  

It is good practice to close the connection when done.

Run the following cell when you are done working on this HW.


In [ ]:
#Close the cursor and connection
conn.close()

---
All done! Good job!